File: typescript/base/src/main/vsdk/toolkit/render/RenderHookList.ts
import { RenderHook } from "./RenderHook";

export class RenderHookList {
  private static renderHookList: RenderHook[] | null = [];

  private constructor() {
  }

  public static renderHooks(): void {
    for (let i = 0; RenderHookList.renderHookList !== null && i < RenderHookList.renderHookList.length; i++) {
      const h = RenderHookList.renderHookList[i]!;
      if (h.function !== null) {
        h.function.apply(h.data);
      }
    }
  }

  public static removeAllRenderHooks(): void {
    RenderHookList.renderHookList = null;
  }
}

File: cpp/base/src/main/vsdk/toolkit/render/RenderHookList.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/render/RenderHookList.h"
#include "vsdk/toolkit/render/RenderHook.h"

java::ArrayList<RenderHook *> *RenderHookList::renderHookList = new java::ArrayList<RenderHook*>();

void
RenderHookList::renderHooks() {
    for ( int i = 0; renderHookList != nullptr && i < renderHookList->size(); i++ ) {
        RenderHook * const h = renderHookList->get(i);
        h->function(h->data);
    }
}

void
RenderHookList::removeAllRenderHooks() {
    delete renderHookList;
    renderHookList = nullptr;
}
